lib/recursive-methods.rb

```diff
@@ -1,49 +1,99 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# source: https://www.bigocheatsheet.com/
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Thank you to Chris-recursion video
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Why can't I use .last?
+def reverse(s, first = 0, last = s.length-1)
```

Comment:
👍  You can't use .last because you won't always using the last index in the string, since you're doing this in place.  You could do it by making a new string with each call, but that's O(n^2) in space complexity.

```ruby
def reverse(s)
  return "" if s.length == 0
  return s.last + s[0..-2]
end
```